shared/src/main/scala/mlscript/ConstraintSolver.scala

```diff
@@ -832,13 +861,53 @@ class ConstraintSolver extends NormalForms { self: Typer =>
             val newBound = (cctx._1 ::: cctx._2.reverse).foldRight(rhs)((c, ty) =>
               if (c.prov is noProv) ty else mkProxy(ty, c.prov))
             lhs.upperBounds ::= newBound // update the bound
+            lhs.tsc.foreachEntry { (tsc, v) =>
+              v.foreach { i =>
+                if (!tsc.tvs(i)._1) {
+                  tsc.updateOn(i, rhs)
+                  if (tsc.constraints.isEmpty) reportError()
+                }
+              }
+            }
+            val u = lhs.tsc.keysIterator.filter(_.constraints.sizeCompare(1)===0).duplicate
+            u._1.foreach { k =>
+              k.tvs.mapValuesIter(_.unwrapProxies).foreach {
+                case (_,tv: TV) => tv.tsc.remove(k)
+                case _ => ()
+              }
+            }
+            u._2.foreach { k =>
+              k.constraints.head.iterator.zip(k.tvs).foreach {
+                case (c, (pol, t)) => if (pol) rec(t, c, false) else rec(c, t, false)
+              }
+            }
```

Comment:
Only execute this part when `noApproximateOverload`. Similar comment for code below.